R/register_extendr.R

```diff
@@ -40,27 +45,38 @@ register_extendr <- function(path = ".", quiet = FALSE, force_wrappers = FALSE,
 
   outfile <- rprojroot::find_package_root_file("R", "extendr-wrappers.R", path = path)
 
-  # If force_wrappers is TRUE, use a minimal wrapper file even when
-  # make_wrappers() fails; since the wrapper generation depends on the compiled
-  # Rust code, the package needs to be installed before attempting this, but
-  # it's not always the case (e.g. the package might be corrupted, or not
-  # installed yet).
-  if (isTRUE(force_wrappers)) {
-    error_handle <- function(e) {
-      cli::cli_alert_danger("Failed to generate wrapper functions: {e$message}.")
-      cli::cli_alert_warning("Falling back to a minimal wrapper file instead.")
-
-      make_example_wrappers(pkg_name, outfile, path = path)
-    }
-  } else {
-    error_handle <- function(e) {
-      ui_throw(
-        "Failed to generate wrapper functions.",
-        c(
-          ui_x(e[["message"]])
-        )
-      )
-    }
+  path <- rprojroot::find_package_root_file(path = path)
+
+  # If compile is NA, compile if the DLL is newer than the source files
+  if (isTRUE(is.na(compile))) {
+    compile <- needs_compilation(path, quiet) || pkgbuild::needs_compile(path)
+  }
+
+  if (isTRUE(compile)) {
+    # This relies on [`pkgbuild::needs_compile()`], which
+    # does not know about Rust files modifications.
+    # `force = TRUE` enforces compilation.
+    pkgbuild::compile_dll(
+      path = path,
+      force = TRUE,
+      quiet = quiet
+    )
+  }
+
+  dll_file <- get_library_path(path)
+
+  # If the wrapper file is newer than the DLL file, assume it's generated by
+  # the latest DLL, which should mean it doesn't need regenerated. This isn't
+  # always the case (e.g. when the user accidentally edited the wrapper file
+  # by hand) so the user might need to run with `force = TRUE`.
+  if (!isTRUE(force) &&
+      isTRUE(file.exists(dll_file)) &&
+      isTRUE(file.exists(outfile)) &&
+      isTRUE(file.info(dll_file)$mtime < file.info(outfile)$mtime)) {
+    cli::cli_alert_info("{.file {outfile}} is up-to-date. Skip generating wrapper functions.")
+    cli::cli_alert_info("If you want to force the generation, run `register_extendr(force = TRUE)`.")
+
+    return(invisible(character(0L)))
   }
```

Comment:
I was talking about `needs_new_wrappers` function that essential compared `mtime`s of the dll and wrapper files.
Now this looks like a little bit of repetition to me.

Would you consider combining this snippet with `needs_compilation()` method, extracting file `mtime` comparison logic into another helper?
E.g.
```R
was_modified_ater <- function(file_paths, compare_to) {
}
```
Here `file_paths` is a paths (character) vector of any length, `compare_to` is a scalar path, to which comparison happens. The output is a logical vector of `length("file_paths")`, returning `TRUE` when `paths$mtime > compare_to$mtime` (pseudo code), `FALSE` otherwise, `NA` if path is invalid or input was `NA`.

This will allow to write something like:
```R
needs_compilation <- function(path, quiet) }
  was_modified_after(get_rust_files(path), get_library_path(path))
  # Verbose message goes here
}
```

And replace this snippet with something like:

```R
was_modified_after(dll_path, out_file)
```

Reply:
Agreed, I'll implement it.

Reply:
I am not sure about naming. If you have a better name than `was_modified_after`, please use it.

Reply:
I chose `find_newer_files_than()`, as it needs to return character vector to show the message here:

https://github.com/extendr/rextendr/blob/def4813dd003efaa9ea7b7b03cbc625c22dcd651/R/track_rust_source.R#L178-L183